### PR TITLE
508 - adding cardinality hint to metadata

### LINKED ISSUE #508: metadata: cardinality hint
At attribute level allow optional definition of a numeric property called `cardinality-hint`. If provided this value will be used by Conseil to determine how to calculate cardinality from live data.

If cardinality hint is less than the low cardinality cut-off (this must be a named constant), hit the table with a `count distinct` query, if it is above the cut-off, simply pass this value as cardinality in the metadata results to avoid counting those field values.

### PR DESCRIPTION
**Pull request by piotrkosecki**
Resolves #508 

Configuration file changes require adding `high-cardinality-limit`:
```
conseil: {
  hostname: "0.0.0.0"
  port: 1337
  cache-ttl: 15 seconds
  max-query-result-size: 100000
+ high-cardinality-limit: 100
}
```

### CODE REVIEW

File: src/main/scala/tech/cryptonomic/conseil/tezos/TezosPlatformDiscoveryOperations.scala [hunk 1]
```diff
@@ -296,22 +292,22 @@ class TezosPlatformDiscoveryOperations(
     * @param  tableName name of the table from which we extract attributes
```
Comment (anonymoussprocket) [inline src/main/scala/tech/cryptonomic/conseil/tezos/TezosPlatformDiscoveryOperations.scala]: doc out of date

File: src/main/scala/tech/cryptonomic/conseil/tezos/TezosPlatformDiscoveryOperations.scala [hunk 2]
```diff
@@ -334,24 +330,30 @@ class TezosPlatformDiscoveryOperations(
...
```
Comment (anonymoussprocket) [inline src/main/scala/tech/cryptonomic/conseil/tezos/TezosPlatformDiscoveryOperations.scala]: doc out of date

File: src/main/scala/tech/cryptonomic/conseil/tezos/TezosPlatformDiscoveryOperations.scala [hunk 3]
```diff
@@ -296,22 +292,22 @@ class TezosPlatformDiscoveryOperations(
...
     * @return list of attributes as a Future
     */
-  override def getTableAttributes(tableName: String): Future[Option[List[Attribute]]] =
+  override def getTableAttributes(entityPath: EntityPath): Future[Option[List[Attribute]]] =
```
Comment (anonymoussprocket) [inline src/main/scala/tech/cryptonomic/conseil/tezos/TezosPlatformDiscoveryOperations.scala]: Why the change from table name to EntityPath

Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/tezos/TezosPlatformDiscoveryOperations.scala]: because extracting metadata configuration for given entity requires full EntityPath, and before there was no information passed about platform/network to get information about given entity

File: src/main/scala/tech/cryptonomic/conseil/tezos/TezosPlatformDiscoveryOperations.scala (hunk 2 above)
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/tezos/TezosPlatformDiscoveryOperations.scala]: changed

File: src/main/scala/tech/cryptonomic/conseil/tezos/TezosPlatformDiscoveryOperations.scala (hunk 1 above)
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/tezos/TezosPlatformDiscoveryOperations.scala]: changed

File: src/main/scala/tech/cryptonomic/conseil/metadata/AttributeValuesCacheConfiguration.scala [hunk 4]
```diff
@@ -16,6 +16,14 @@ class AttributeValuesCacheConfiguration(metadataConfiguration: MetadataConfigura
         .flatMap(_.cacheConfig)
     }.flatten
 
+  /** extracts cache configuration for given attribute path */
+  def getCardinalityHint(path: AttributePath): Option[Int] =
+    when(metadataConfiguration.isVisible(path)) {
```
Comment (ivanopagano) [inline src/main/scala/tech/cryptonomic/conseil/metadata/AttributeValuesCacheConfiguration.scala]: What about introducing a `whenVisible(path)(optionResult)` local method to lighten up these calls?

File: src/main/scala/tech/cryptonomic/conseil/tezos/TezosPlatformDiscoveryOperations.scala [hunk 5]
```diff
@@ -331,27 +327,33 @@ class TezosPlatformDiscoveryOperations(
...
   /** Query for returning partial attributes with updated counts */
-  private def getUpdatedAttributesQuery(tableName: String, columns: List[Attribute]): DBIO[List[Attribute]] =
+  private def getUpdatedAttributesQuery(entityPath: EntityPath, columns: List[Attribute]): DBIO[List[Attribute]] =
     DBIOAction.sequence {
       columns.map { column =>
-        if (canQueryType(column.dataType) && isLowCardinality(column.cardinality)) {
-          TezosDatabaseOperations.countDistinct(tableName, column.name).map { count =>
+        val attributePath = AttributePath(column.name, entityPath)
+        val cardinalityHint = cacheOverrides.getCardinalityHint(attributePath)
+        println(s"$column : $cardinalityHint")
```
Comment (ivanopagano) [inline src/main/scala/tech/cryptonomic/conseil/tezos/TezosPlatformDiscoveryOperations.scala]: leftover?

File: src/main/scala/tech/cryptonomic/conseil/metadata/AttributeValuesCacheConfiguration.scala (hunk 4 above)
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/metadata/AttributeValuesCacheConfiguration.scala]: good idea, added

File: src/main/scala/tech/cryptonomic/conseil/tezos/TezosPlatformDiscoveryOperations.scala (hunk 5 above)
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/tezos/TezosPlatformDiscoveryOperations.scala]: yes, removed

File: src/main/resources/reference.conf [hunk 6]
```diff
@@ -8,6 +8,7 @@ conseil: {
   port: 1337
   cache-ttl: 15 minutes
   max-query-result-size: 100000
+  high-cardinality-limit: 100
```
Comment (anonymoussprocket) [inline src/main/resources/reference.conf]: If this is missing from the config file, how will the application react?

Comment (piotrkosecki) [reply src/main/resources/reference.conf]: conseil will not start because of missing value in the config
